apps/service_providers/llm_service/prompt_context.py

```diff
@@ -134,3 +136,66 @@ def __eq__(self, other):
 
 
 EMPTY = SafeAccessWrapper(None)
+
+
+class ParticipantDataProxy:
+    """Allows multiple access without needing to re-fetch from the DB"""
+
+    @classmethod
+    def from_state(cls, pipeline_state) -> Self:
+        # using `.get` here for the sake of tests. In practice the session should always be present
+        return cls(pipeline_state.get("experiment_session"))
+
+    def __init__(self, experiment_session):
+        self.session = experiment_session
+        self._participant_data = None
+
+    def _get_db_object(self):
+        if not self._participant_data:
+            self._participant_data, _ = ParticipantData.objects.get_or_create(
+                participant_id=self.session.participant_id,
+                experiment_id=self.session.experiment_id,
+                team_id=self.session.experiment.team_id,
+            )
+        return self._participant_data
+
+    def get(self):
+        data = self._get_db_object().data
+        return self.session.participant.global_data | data
+
+    def set(self, data):
+        participant_data = self._get_db_object()
+        participant_data.data = data
+        participant_data.save(update_fields=["data"])
+
+        self.session.participant.update_name_from_data(data)
+
+    def get_schedules(self):
```

Comment:
I think this should be cached

Reply:
I should have been clearer. We can't use this cache decorator here since it will create a global cache which will result in a memory leak. Options are to create a `cached_property` which is called by this function or create a private variable which you set once on first call.